File: reactor-java/src/test/java/com/github/mostroverkhov/r2/java/RequesterAdapterTest.java
package com.github.mostroverkhov.r2.java;

import com.github.mostroverkhov.r2.core.internal.requester.CallAdapter;
import com.github.mostroverkhov.r2.java.adapters.JavaRequesterAdapter;
import io.rsocket.AbstractRSocket;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.lang.reflect.Method;
import java.time.Duration;
import java.util.function.Function;

public class RequesterAdapterTest {

    private CallAdapter adapter;

    @Before
    public void setUp() {
        adapter = new JavaRequesterAdapter(new AbstractRSocket() {
        });
    }


    @Test(expected = RuntimeException.class)
    public void wrongReturnType() throws Exception {
        Method obj = Stub.class.getMethod("obj");
        adapter.resolve(obj, new RuntimeException());
    }

    @Test
    public void monoType() throws Exception {
        assertError("mono", Mono.class, Mono::flux);
    }

    @Test
    public void fluxType() throws Exception {
        assertError("flux", Flux.class, flux -> flux);
    }

    private <T> void assertError(String method,
                                 Class<T> resolvedType,
                                 Function<T, Flux<?>> toFlux) throws Exception {
        Method obj = Stub.class.getMethod(method);
        RuntimeException ex = new RuntimeException();
        Object resolved = adapter.resolve(obj, ex);
        Assert.assertTrue(resolvedType.isAssignableFrom(resolved.getClass()));
        T cast = resolvedType.cast(resolved);
        Flux<?> compl = toFlux.apply(cast);
        StepVerifier.create(compl).expectNextCount(0)
                .expectError(ex.getClass())
                .verify(Duration.ofSeconds(1));
    }

    private interface Stub {
        Object obj();

        Mono<String> mono();

        Flux<String> flux();
    }
}

File: rxjava-kotlin/src/test/java/com/github/mostroverkhov/r2/android/RequesterAdapterTest.kt
package com.github.mostroverkhov.r2.android

import com.github.mostroverkhov.r2.android.adapters.RequesterAdapter
import com.github.mostroverkhov.r2.core.internal.requester.CallAdapter
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.subscribers.TestSubscriber
import io.rsocket.android.AbstractRSocket
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class RequesterAdapterTest {
    lateinit var adapter: CallAdapter
    @Before
    fun setUp() {
        adapter = RequesterAdapter(object : AbstractRSocket() {})
    }


    @Test(expected = RuntimeException::class)
    fun wrongReturnType() {
        val obj = Stub::class.java.getMethod("obj")
        adapter.resolve(obj, RuntimeException())
    }

    @Test
    fun singleType() {
        assertError<Single<Any>>("single", { it.toFlowable() })
    }

    @Test
    fun flowableType() {
        assertError<Flowable<Any>>("flowable", { it })
    }

    @Test
    fun completableType() {
        assertError<Completable>("compl", { it.toFlowable() })
    }

    private inline fun <reified T> assertError(method: String,
                                               toFlowable: (T) -> Flowable<Any>) {

        val obj = Stub::class.java.getMethod(method)
        val ex = RuntimeException()
        val resolve = adapter.resolve(obj, ex)
        assertTrue(resolve is T )
        val compl = toFlowable(resolve as T)
        val subscriber = TestSubscriber<Any>()
        compl.blockingSubscribe(subscriber)
        subscriber.assertNoValues()
        subscriber.assertError(ex.javaClass)
    }

    private interface Stub {
        fun obj(): Any

        fun single(): Single<String>

        fun flowable(): Flowable<String>

        fun compl(): Completable
    }
}